plugin/documents.py

```diff
@@ -959,6 +965,32 @@ def _update_stored_selection_async(self) -> Tuple[Optional[sublime.Region], bool
         self._stored_selection = selection
         return changed_first_region, True
 
+    def _format_on_paste_async(self) -> None:
+        self.purge_changes_async()
+        clipboard_text = sublime.get_clipboard()
+        sel = self.view.sel()
+        split_clipboard_text = clipboard_text.split('\n')
+        multi_cursor_paste = len(split_clipboard_text) == len(sel) and len(sel) > 1
+        original_selection = list(sel)
+        regions_to_format = []  # type: List[sublime.Region]
+        pasted_text = clipboard_text
+        # add regions to selection, in order for lsp_format_document_range to format those regions
+        for index, region in enumerate(sel):
+            if multi_cursor_paste:
+                pasted_text = split_clipboard_text[index]
+            pasted_region = self.view.find(pasted_text, region.end(), sublime.REVERSE | sublime.LITERAL)
+            if pasted_region:
+                # Including whitespace may help servers format a range better
+                # More info at https://github.com/sublimelsp/LSP/pull/2311#issuecomment-1688593038
+                a = self.view.find_by_class(pasted_region.a, False,
+                                            sublime.CLASS_WORD_END | sublime.CLASS_PUNCTUATION_END)
```

Comment:
@jwortmann your original code snippet contained these flags `sublime.CLASS_WORD_END | sublime.CLASS_PUNCTUATION_END | sublime.CLASS_LINE_START`. As you can see above `sublime.CLASS_LINE_START` was left out. 

If `sublime.CLASS_LINE_START` was specified, then LSP-json would not format a region as expected.